### ISSUE TITLE
Speculative decoding: turn-final stop token is never committed, so multi-turn transcripts lose <|im_end|> after every assistant turn

### ISSUE THREAD

**Issue by john-rocky** [google-ai-edge/LiteRT-LM#3439]
Thanks for shipping speculative decoding in 0.16.0 — the drafter path works end to end on a bundle I exported myself (Qwen3.5 + a 1-layer MTP drafter), and single-turn decode A/Bs cleanly against the same bundle with the flag off.

While gating multi-turn conversations under `--speculative-decoding true` I hit a structural issue: **the stop token that ends each assistant turn is never committed to the KV cache or the step counter, and the next turn's prefill does not re-add it.** Every assistant turn in a multi-turn session is therefore missing its `<|im_end|>` in the context the model actually sees. A 2B model mostly shrugs it off; a 0.8B degrades visibly on later turns (echoed words, replies truncated to their first token).

Measured with the released 0.16.0 python API on the CPU backend (`enable_benchmark=True`, reading `last_decode_token_count`, `last_prefill_token_count` and `token_count` after each turn; greedy, same 5-turn word-game conversation both legs; reply "MANGO" = 2 tokens + stop):

| leg | committed per turn | next-turn prefill | KV total drift |
|---|---|---|---|
| flag off | 3 (reply + stop) | 21 | baseline |
| flag on | 2 (reply only) | 21 (identical) | −1 per turn |

Since the prefill increments are identical in both legs, the missing token is not compensated anywhere — the flag-on transcript is `…MANGO\n<|im_start|>user…` where the flag-off one is `…MANGO<|im_end|>\n<|im_start|>user…`.

Reading the OSS source, this looks consistent with `LlmLiteRtCompiledModelExecutorBase::Decode`'s MTP branch: it advances `current_step` by the size of the token vector returned from `MtpDrafter::Draft()`, and neither that path nor the next prefill appears to handle a stop token that arrives as the round's bonus token (the block reads the same to me at the 0.16.0 tag and at main 90f42140) — though you know this code far better than I do, so please treat that as an observation rather than a diagnosis.

The probe is ~30 lines against the public python API and reproduces on any MTP-capable bundle; happy to share the exact script, or the Qwen3.5 test bundle, if either is useful. Single-turn generation and benchmarks are unaffected, so this only matters once conversations run multi-turn under the flag.

Thanks again for the MTP runtime — with this one exception it has been a pleasure to build against.

**Comment by nireekshak** [comment]
Hi @john-rocky ,

Thank you for the detailed report, clear breakdown, and source pointer. Could you please share your ~30-line reproduction script and the Qwen3.5 MTP bundle (or the export command/recipe used to generate it) to help us investigate this issue further?

**Comment by john-rocky** [comment]
Thanks for picking this up, @nireekshak. Script first, then the bundle.

**Reproduction script.** Re-run today against litert-lm 0.16.0 from pip (CPU backend, macOS on an Apple M4 Max). One argument, the `.litertlm` path:

```python
#!/usr/bin/env python3
"""Does speculative decoding commit the turn-final stop token?

Runs the same 5-turn greedy conversation with the flag off and on and prints,
after every turn, the engine's committed decode-token count, the size of that
turn's prefill, and the conversation's total token_count.
litert-lm 0.16.0 python API, CPU backend.  Usage: python3 probe.py <bundle>"""
import sys
import litert_lm

TURNS = [
    "Reply with only the single word MANGO.",
    "Reply with only the single word KIWI.",
    "What was the first word you replied with in this conversation? "
    "Reply with only that word.",
    "Repeat your previous reply exactly.",
    "Reply with only the single word LEMON.",
]

def text_of(stream):
    return "".join(i.get("text", "") for c in stream
                   for i in c.get("content", []) if i.get("type") == "text")

for spec in (False, True):
    print(f"--- enable_speculative_decoding={spec}", flush=True)
    eng = litert_lm.Engine(sys.argv[1], backend=litert_lm.Backend.CPU(),
                           enable_speculative_decoding=spec,
                           max_num_tokens=2048, enable_benchmark=True)
    sc = litert_lm.SamplerConfig(top_k=1, temperature=0.0, seed=0)
    with eng as engine, engine.create_conversation(sampler_config=sc) as conv:
        for prompt in TURNS:
            text = text_of(conv.send_message_async(prompt))
            info = conv.get_benchmark_info()
            print(f"{text[:20]!r:24} decode_committed={info.last_decode_token_count:<3}"
                  f" prefill={info.last_prefill_token_count:<3}"
                  f" token_count={conv.token_count}", flush=True)
```

Output on my Qwen3.5-0.8B MTP bundle:

```
--- enable_speculative_decoding=False
'MANGO'                  decode_committed=3   prefill=21  token_count=24
'KIWI'                   decode_committed=3   prefill=21  token_count=48
'MANGO'                  decode_committed=3   prefill=30  token_count=81
'MANGO'                  decode_committed=3   prefill=18  token_count=102
'MANGO'                  decode_committed=3   prefill=21  token_count=126
--- enable_speculative_decoding=True
'MANGO'                  decode_committed=2   prefill=21  token_count=23
'MANGO'                  decode_committed=2   prefill=21  token_count=46
'MANGO'                  decode_committed=2   prefill=30  token_count=78
'M'                      decode_committed=1   prefill=18  token_count=97
'M'                      decode_committed=1   prefill=21  token_count=119
```

The prefill sizes are identical in both legs, so the token missing from each decode is never re-added, and `token_count` drifts by one per turn (by two on the turns where the 0.8B already truncates to a single token). My 2B bundle shows the same drift with the text intact: `token_count` 24/48/81/102/127 flag-off vs 23/46/78/98/121 flag-on. The accounting is the same on both; only how far the model copes differs.

**The bundle.** It comes from my own export path, so I cannot give you a one-line command on released tooling. The recipe, in three steps:

- Base graph: `Qwen/Qwen3.5-0.8B` through litert-torch `export_hf` at commit 115a136 plus a local patch of mine (ring-addressed gated-delta state so verify rounds can rewind, a `verify` signature with per-position logits, `activations` on decode), with `externalize_embedder=True`, cache length 4096, three draft steps.
- Drafter: a separately exported 1-layer graph (the checkpoint's `mtp.*` tensors plus the tied head), packed as a `tf_lite_mtp_drafter` section with the litert-lm 0.16.0 `unpack`/`pack` CLI (`model_type = "mtp_drafter"` appended to model.toml).
- Weights: dynamic-range int8 on the linears and embedding tables of both tflites (ai-edge-quantizer 0.9.0).

Nothing in the script is model-specific, so I would expect any bundle with a drafter section to show the same two-versus-three count. If a ready-made file is easier, I can upload the 0.8B bundle (1.47 GB) to Hugging Face together with the patch and driver scripts — entirely your call.

Thanks again for looking into it.

**Comment by john-rocky** [comment]
Follow-up on the bundle: rather than leave it as an offer, I uploaded the 0.8B one so the script can be run as-is.

- Bundle + the probe script: https://huggingface.co/mlboydaisuke/Qwen3.5-0.8B-MTP-repro-LiteRT (`Qwen3.5-0.8B_mtp_drafter_int8.litertlm`, 1.47 GB; the card carries the expected probe output).
- Patch + driver scripts + gate record: https://github.com/john-rocky/hf-to-litertlm/tree/main/mtp_work (the three-step recipe from my previous comment, runnable).

One addition since yesterday: the probe gives the same numbers on litert-lm 0.16.1 (pip, CPU, macOS), so both released versions show it.

Thanks again.

**Comment by nireekshak** [comment]
Hi @john-rocky ,

Thanks for providing the reproduction probe script, the export recipe, and uploading the ready-to-test model bundle to Hugging Face! We will investigate this further using your provided artifacts and get back to you.
